ops/testing.py

```diff
@@ -0,0 +1,408 @@
+        :rtype: dict
+        :raises: KeyError if relation_id doesn't exist
+        """
+        return self._backend._relation_data[relation_id].get(app_or_unit, None)
+
+    def update_relation_data(self, relation_id, app_or_unit, key_values):
+        """Update the relation data for a given unit or application in a given relation.
+
+        This also triggers the `relation_changed` event for this relation_id.
+
+        :param relation_id: The integer relation_id representing this relation.
+        :param app_or_unit: The unit or application name that is being updated.
+          This can be the local or remote application.
+        :param key_values: Each key/value will be updated in the relation data.
+        :return: None
+        """
+        new_values = self._backend._relation_data[relation_id][app_or_unit].copy()
+        for k, v in key_values.items():
+            if v == '':
+                new_values.pop(k, None)
+            else:
+                new_values[k] = v
+        self._backend._relation_data[relation_id][app_or_unit] = new_values
+        relation_name = self._backend._relation_names[relation_id]
+        if self._model is not None:
+            relation = self._model.get_relation(relation_name, relation_id)
+            if '/' in app_or_unit:
+                entity = self._model.get_unit(app_or_unit)
+            else:
+                entity = self._model.get_app(app_or_unit)
+            rel_data = relation.data.get(entity, None)
+            if rel_data is not None:
+                # If we have read and cached this data, make sure we invalidate it
+                rel_data._invalidate()
+        # TODO: we only need to trigger relation_changed if it is a remote app or unit
+        self._emit_relation_changed(relation_id, app_or_unit)
+
+    def _emit_relation_changed(self, relation_id, app_or_unit):
+        if self._charm is None or not self._hooks_enabled:
+            return
+        rel_name = self._backend._relation_names[relation_id]
+        relation = self.model.get_relation(rel_name, relation_id)
+        if '/' in app_or_unit:
+            app_name = app_or_unit.split('/')[0]
+            unit_name = app_or_unit
+            app = self.model.get_app(app_name)
+            unit = self.model.get_unit(unit_name)
+            args = (relation, app, unit)
+        else:
+            app_name = app_or_unit
+            app = self.model.get_app(app_name)
+            args = (relation, app)
+        self._charm.on[rel_name].relation_changed.emit(*args)
+
+    def update_config(self, key_values=None, unset=()):
+        """Update the config as seen by the charm, and trigger a config_changed event.
+
+        This will trigger a `config_changed` event.
+
+        :param key_values: A Mapping of key:value pairs to update in config.
+        :param unset: An iterable of keys to remove from Config. (Note that this does
+          not currently reset the config values to the default defined in config.yaml.
```

Comment:
```suggestion
          not currently reset the config values to the default defined in config.yaml)
```
(or something like that -- close that paren!)